### ISSUE TITLE
Revoked signal gets emitted at execution time

### ISSUE THREAD

**Issue by pyahmed** [coleifer/huey#478]
When revoking a scheduled task, the `SIGNAL_REVOKED` is emitted at the originally scheduled time (or close to it) but not immediately at revocation.

Is this the intended behaviour? For my usecase it would be useful to have the signal emitted immediately when a task has been revoked (canceled). The alternative is to do constant polling or implement a different notification scheme than for `SIGNAL_SCHEDULED`, `SIGNAL_COMPLETE` etc.

/EDIT I also realized that if you restore the task befor the scheduled execution time you will not be informed that it has been "scheduled" again. So I think it would be good to have `SIGNAL_RESTORED` or `SIGNAL_SCHEDULED` for those tasks.

**Comment by coleifer** [comment]
[Docs](https://huey.readthedocs.io/en/latest/signals.html):

> SIGNAL_REVOKED: task is revoked and will not be executed.

When the consumer receives a task, but that task has been revoked, the SIGNAL_REVOKED will be fired - that is the intended behavior.

For your use-case -- since you are the one revoking the task, presumably you are able to run any additional code at that time. There should be no need to "poll" or do anything crazy. When you revoke the task, just run whatever additional logic you wish. There is no restored signal because restoring a task is a user-defined operation.

**Comment by pyahmed** [comment]
Well what I meant is that this way one would need two types of logic:
- handle scheduled/complete via signals
- handle revocation/restoration via custom code

It would be much easier and elegant if all notification logic could be implemented via signals in one place.

IMHO getting the revocation signal at the execution time is not very useful but I might be looking at it the wrong way.

Thanks for the reply.

**Comment by coleifer** [comment]
You already know when a task revocation takes place **because you're the one who's revoking it**. The signals allow you to respond to certain events that are not in your control -- like when the consumer picks up a task to start working on.

**Comment by pyahmed** [comment]
By that same logic I already know when I schedule a task, yet there is a signal emitted for it at the time of scheduling.

When I schedule a task, revoke it multiple times to change ('update') the arguments and then schedule it again (same time) then I end up with multiple 'revoked - not executing' in the log file at the same time as the actual task is executed. This way I won't know based on the huey consumer log when the task was actually revoked. Getting the signal at execution time has no benefit or at least I don't see any in it.

**Comment by coleifer** [comment]
>By that same logic I already know when I schedule a task, yet there is a signal emitted for it at the time of scheduling.

But you don't necessarily know when a task will be executed / picked-up by the consumer. The signals provide visibility into a process that you do not control - the consumer. They let you respond to things the consumer is doing on it's own.

> This way I won't know based on the huey consumer log when the task was actually revoked.

You are revoking it. You know when the task was actually revoked.

**Comment by pyahmed** [comment]
> But you don't necessarily know when a task will be executed / picked-up by the consumer. 

I wrote scheduling not execution or is picked-up by consumer == scheduled?

Oh well it would be nice to have this behaviour or as an option to see it in the consumer log when a task was revoked.

**Comment by coleifer** [comment]
I've been trying, with apparently no success, to impress upon you the fact that *you* know when a task was revoked. The consumer does not - not until it picks up the task for execution.

**Comment by pyahmed** [comment]
> I've been trying, with apparently no success, to impress upon you the fact that you know when a task was revoked. The consumer does not - not until it picks up the task for execution.

Alright so when I schedule a task; the consumer immediately logs the fact that a task has been scheduled. Do I understand correctly that the huey consumer (==huey_consumer.py) does not **receive** this kind information in case of a revocation and if that is true when does it get/receive this info? Only upon reading the task details from the queue? Is this by just 'marking' a task as revoked? 

Then.... (hihi) it would be nice to put a field to the task object that contains the revocation date/time :-)

**Comment by coleifer** [comment]
>Alright so when I schedule a task; the consumer immediately logs the fact that a task has been scheduled

When you schedule a task, the task goes into the queue and the consumer picks it up. It sees that it is not yet time to run it, so it puts it on the schedule (at which point the scheduler enqueues it at the appropriate time).

>does not receive this kind information in case of a revocation and if that is true when does it get/receive this info?

Correct. It checks revocation at the time the task is going to be executed.
